### ISSUE TITLE
Allow passing authentication details through environment variables

### ISSUE THREAD

**Issue by malletgu** [prometheus-community/elasticsearch_exporter#307]
Hi, thanks for this exporter :) 

With the announce of the ES operator by elastic we started to use ES on k8s and quickly ran into an issue with passing generated credentials to the exporter.
As Kubernetes doesn't allow injection of secrets within string and credentials must be passed in the URI.

I created a pull request where I added 2 arguments for passing the credentials as environments variables. This is currently working well for us using https://github.com/elastic/cloud-on-k8s.

**Comment by mfouilleul** [comment]
Why there is no answer on this issue? It seems critical no?

Thanks,

Max.

**Comment by zwopir** [comment]
sorry for the late reply, this is after all an open source project and I'm not working on it full-time. Still, I try to create a new release next week.

Regarding the issue: I don't consider this one critical. I you need to hide the credentials, you can put the whole URL into a k8s secret.

But the PR is still appreciated and I'm truly sorry the release cycle is not anywhere near being perfect.

**Comment by mfouilleul** [comment]
Hum, yes you'r right I missed the default value set to ES_URI, sorry. 

Thanks.

**Comment by malletgu** [comment]
The issue I had with ES_URI is that the operator provided by elastic contains the user in a secret and it is impossible currently to inject a secret in the URI on kubernetes so I had to use environment variables to avoid hard coding the URI :/ 
Please see my pull request if it is good enough for you :) https://github.com/justwatchcom/elasticsearch_exporter/pull/308

**Comment by mfouilleul** [comment]
Also the problem is, in the chart elasticsearch_exporter there is a default value for uri, making impossible to use the default ES_URI env var...
https://github.com/helm/charts/blob/master/stable/elasticsearch-exporter/values.yaml#L73

**Comment by mfouilleul** [comment]
Without this, we would be able to inject a secret to set ES_URI + use this secret with the chart variable es:envFromSecret.

**Comment by vosdev** [comment]
I would want this feature to not show my credentials in my process list

```
_elasti+  66522  0.0  0.3 112952 13624 ?        Ssl  14:16   0:00 /usr/local/bin/elasticsearch_exporter --es.uri=https://user:password@localhost:9200 --es.timeout=5s --es.clusterinfo.interval=5m --web.telemetry-path=/metrics --web.listen-address=:9114
```

**Comment by malletgu** [comment]
Hi @vosdev have you seen this behaviour while using environment variables ? It shouldn't show the credentials in the process lists are the credentials won't be part of the process args.

**Comment by vosdev** [comment]
> Hi @vosdev have you seen this behaviour while using environment variables ? It shouldn't show the credentials in the process lists are the credentials won't be part of the process args.

Does the elasticsearch_exporter already support the usage of environment variables for username:password ? At the moment I have to put those in the URL for basic auth like https://username:password@es1.example.com. I would like to not have to do that. I was under the impression that is what this Issue is for :)

**Comment by mmcgarr** [comment]
If it helps anyone, setting es.uri to an empty string in your values.yaml means the ES_URI environment variable will not be overridden:
```
es:
  uri: ''
```

**Comment by schrej** [comment]
I've also solved this by setting the following on the chart.
```yaml
env:
  ES_URI: "https://$(ES_USER):$(ES_PASSWORD)@elasticsearch:9200"
es:
  uri: null
```
Both `ES_USER` and `ES_PASSWORD` are stored in a secret.

**Comment by malletgu** [comment]
Thanks @schrej I wasn't aware of this, I'll test this then close this MR :)

**Comment by sysadmind** [comment]
@schrej could you provide a working example? I have been trying to replicate your results with no success. Here's a minimal example that I tried, but was unsuccessful with getting the environment variables to work. I still think that #308 is necessary in order to get this working for the ECK use case w/ the stable/elasticsearch-exporter helm chart.

On a fresh minikube I applied ECK v1.1.0
`kubectl apply -f https://download.elastic.co/downloads/eck/1.1.0/all-in-one.yaml`

Then I deployed a new ECK elasticsearch cluster:
```
cat <<EOF | kubectl apply -f -
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: quickstart
spec:
  version: 7.7.0
  nodeSets:
  - name: default
    count: 1
    config:
      node.master: true
      node.data: true
      node.ingest: true
      node.store.allow_mmap: false
    podTemplate:
      spec:
        volumes:
        - name: elasticsearch-data
          emptyDir: {}
EOF
```

Then I created a helm chart overrides file:
```yaml
#exporter-override.yaml
es:
  uri: null
  sslSkipVerify: true

extraEnvSecrets:
  ES_PASSWORD:
    secret: quickstart-es-elastic-user
    key: elastic

env:
  ES_USER: elastic
  ES_URI: "https://$(ES_USER):$(ES_PASSWORD)@quickstart-es-http:9200"
```

Next I installed the helm chart:
`helm install quickstart-exporter stable/elasticsearch-exporter --version 3.3.0 -f exporter-override.yaml`

Afterwards, the pods are running and ready:
```
NAME                                                          READY   STATUS    RESTARTS   AGE
quickstart-es-default-0                                       1/1     Running   0          4m28s
quickstart-exporter-elasticsearch-exporter-5484df96fc-294cg   1/1     Running   0          3m35s
```
Scraping the metrics endpoint for the exporter fails with this in the logs:
```
level=warn ts=2020-05-14T19:31:07.410243514Z caller=snapshots.go:265 msg="failed to fetch and decode snapshot stats" err="HTTP Request failed with code 401"
level=warn ts=2020-05-14T19:31:07.421632563Z caller=nodes.go:1851 msg="failed to fetch and decode node stats" err="HTTP Request failed with code 401"
level=warn ts=2020-05-14T19:31:07.421860452Z caller=indices_settings.go:116 msg="failed to fetch and decode cluster settings stats" err="HTTP Request failed with code 401"
level=warn ts=2020-05-14T19:31:07.431273092Z caller=cluster_health.go:270 msg="failed to fetch and decode cluster health" err="HTTP Request failed with code 401"
level=warn ts=2020-05-14T19:31:07.4355574Z caller=indices.go:1061 msg="failed to fetch and decode index stats" err="HTTP Request failed with code 401"
```
And I checked the environment variables which are set, but there does not seem to be substitution with the method you describe. Maybe this is a limitation of the helm chart?
```
kubectl exec -it quickstart-exporter-elasticsearch-exporter-5484df96fc-294cg -- sh
/ $ echo $ES_USER
elastic
/ $ echo $ES_PASSWORD
o61B8O5hZE09URt50jCPCP50
/ $ echo $ES_URI
https://$(ES_USER):$(ES_PASSWORD)@quickstart-es-http:9200
```

My versions:
```
$ kubectl version --short
Client Version: v1.18.2
Server Version: v1.18.2

$ minikube version
minikube version: v1.10.1
commit: 63ab801ac27e5742ae442ce36dff7877dcccb278
```

**Comment by schrej** [comment]
According to [documentation](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#envvar-v1-core) the order of the variables matters.
I was using `envFromSecret` which is mapped to k8s's `envFrom` property. `extraEnvSecrets` is applied to k8s's `env` after the `env` property of the helm chart.
I assume `envFrom` is handled by k8s before `env` is.
The proper fix for this would probably be to switch [this](https://github.com/helm/charts/blob/master/stable/elasticsearch-exporter/templates/deployment.yaml#L52-L55) and [this](https://github.com/helm/charts/blob/master/stable/elasticsearch-exporter/templates/deployment.yaml#L56-L63).

**Comment by kalinon** [comment]
Heres my working example using an elasticsearch operator deployment:
```yaml
secretMounts:
  - name: es-certs
    secretName: elasticsearch-es-http-certs-internal
    path: /ssl

envFromSecret: elasticsearch-es-elastic-user

env:
  ES_URI: "https://elastic:$(elastic)@elasticsearch-es-http:9200"

es:
  snapshots: false
  uri: null
  ## Skip SSL verification when connecting to Elasticsearch
  sslSkipVerify: false
  ssl:
    ## If true, a secure connection to ES cluster is used
    enabled: true
    ## If true, certs from secretMounts will be need to be referenced instead of certs below
    useExistingSecrets: true
    ca:
      # Path of ca pem file which should match a secretMount path
      path: /ssl/ca.crt
    client:
      enabled: false
serviceMonitor:
  enabled: true
  namespace: prometheus
  labels:
    app: prometheus
    function: metrics
```

**Comment by creckord** [comment]
I still think that #308 should be merged, because if I'm not mistaken this solution only works with Helm, while the merge request would also solve this for plain K8s.

**Comment by schrej** [comment]
The environment variable nesting is a k8s feature. So it will work with k8s. That PR might make it easier in some cases though.

**Comment by sastorsl** [comment]
For anyone else running through docker setting in environment variables solves the immediate issue of displaying credentions through `ps -ef`.

      elk_d1_elasticsearch_client-exporter:
        image: <imageurl>/justwatch/elasticsearch_exporter:1.1.0
        command: '--es.uri=https://$$ES_USER:$$ES_PW@elasticsearcht:9200 --es.ca=/certs/ca/ca.crt --es.all --es.indices --es.shards'
        restart: always
        environment:
          - ES_USER=elastic
          - ES_PW=mypass

**Comment by nickfan** [comment]
workaround:

```yaml
version: '3.6'
x-logging:
  &default-logging
  driver: json-file
  options:
    max-size: "10M"
    max-file: "10"
networks:
  app-network:
    driver: ${NETWORKS_DRIVER}
    external: true
    name: ${NETWORKS_ALIAS}
services:
  elasticsearch_exporter:
    env_file: [ '../../.env', './.env.app' ]
    logging: *default-logging
    restart: unless-stopped
    networks:
      - app-network
    container_name: elasticsearch_exporter
    image: quay.io/prometheuscommunity/elasticsearch-exporter:latest
    entrypoint: ["/bin/sh", "-c", "/bin/elasticsearch_exporter --es.uri=$${PM_EXPORTER_ES_URI}"]
#    command: [ "--es.uri=$${PM_EXPORTER_ES_URI}"] # not work
    ports:
      - '9114:9114'

```

**Comment by sysadmind** [comment]
This functionality was resolved in #461 and should be usable in the next release
